--- postgrey-1.33/postgrey	2010-05-04 22:51:52.000000000 +0200
+++ postgrey	2010-06-21 13:56:29.011291262 +0200
@@ -141,6 +141,31 @@
     $self->{postgrey}{whitelist_recipients} = \@whitelist_recipients;
 }
 
+
+sub read_reverse_hostname_regexps($) {
+   my ($self) = @_;
+   my @z = ();
+   my @l = ();
+
+   if(open(F, $self->{postgrey}{zombie_regexp_file})) {
+      while(<F>) {
+         if(/^\#/ || /^$/) { next; }
+
+         chomp;
+         @l = split(/\s/, $_);
+
+         if($l[0] =~ /^\/(\S+)\/$/) {
+            push @z, qr{$1}i;
+         }
+
+      }
+      close F;
+   }
+
+   $self->{postgrey}{zombie_regexps} = \@z;
+}
+
+
 sub do_sender_substitutions($$)
 {
     my ($self, $addr) = @_;
@@ -339,10 +364,31 @@
 {
     my ($self, $now, $attr) = @_;
     my $db  = $self->{postgrey}{db};
+    my $zombie = 0;
+
     # This attribute occurs in connections from the policy-test script,
     # not in a regular postfix query connection
     if(defined $attr->{policy_test_time}) { $now = $attr->{policy_test_time} }
 
+
+    for my $z (@{$self->{postgrey}{zombie_regexps}}) {
+       if($attr->{client_name} =~ $z) {
+          $zombie = 1;
+       }
+    }
+
+    if ($zombie == 0) {
+       $self->mylog_action($attr, 'pass', 'unlikely a zombie');
+       return 'DUNNO';
+    }
+
+
+    if ($self->{postgrey}{dryrun}) {
+       $self->mylog(1, "would GREYLIST " . $attr->{client_name});
+       return 'DUNNO';
+    }
+
+
     # whitelists
     for my $w (@{$self->{postgrey}{whitelist_clients}}) {
         if($attr->{client_name} =~ $w) {
@@ -511,6 +557,7 @@
         'whitelist-clients=s@', 'whitelist-recipients=s@',
         'syslogfacility|syslog-facility|facility=s',
         'retry-window=s', 'greylist-action=s', 'greylist-text=s', 'privacy',
+        'dryrun',
         'hostname=s', 'exim', 'listen-queue-size=i', 'x-greylist-header=s',
     ) or exit(1);
     # note: lookup-by-subnet can be given for compatibility, but it is default
@@ -609,7 +656,9 @@
                   "$CONFIG_DIR/postgrey_whitelist_clients.local" ],
             whitelist_recipients_files => $opt{'whitelist-recipients'} ||
                 [ "$CONFIG_DIR/postgrey_whitelist_recipients" ],
+            zombie_regexp_file => "$CONFIG_DIR/zombie.regexp",
             privacy => defined $opt{'privacy'},
+            dryrun => defined $opt{'dryrun'},
             hostname => defined $opt{hostname} ? $opt{hostname} : hostname,
             exim => defined $opt{'exim'},
             x_greylist_header  => $opt{'x-greylist-header'} || 'X-Greylist: delayed %t seconds by postgrey-%v at %h; %d',
@@ -622,6 +671,7 @@
     # read whitelist
     $server->read_clients_whitelists();
     $server->read_recipients_whitelists();
+    $server->read_reverse_hostname_regexps();
 
     # --privacy requires Digest::SHA1
     if($opt{'privacy'}) {
@@ -644,6 +694,7 @@
     $self->mylog(1, "HUP received: reloading whitelists...");
     $self->read_clients_whitelists();
     $self->read_recipients_whitelists();
+    $self->read_reverse_hostname_regexps();
 }
 
 sub post_bind_hook()
@@ -800,6 +851,7 @@
      --greylist-action=A if greylisted, return A to Postfix (default: DEFER_IF_PERMIT)
      --greylist-text=TXT response when a mail is greylisted
                          (default: Greylisted + help url, see below)
+     --dryrun            not greylisting actually, just log what we would do
      --lookup-by-subnet  strip the last 8 bits from IP addresses (default)
      --lookup-by-host    do not strip the last 8 bits from IP addresses
      --privacy           store data using one-way hash functions
