--- mailgraph.orig	2011-04-23 21:15:50.000000000 +0200
+++ mailgraph	2011-06-23 11:25:29.468582144 +0200
@@ -632,6 +632,11 @@
 				event($time, 'rejected');
 			}
 		}
+		elsif($prog eq 'postscreen') {
+			if($text =~ /DISCONNECT/) {
+				event($time, 'rejected');
+			}
+		}
 	}
 	elsif($prog eq 'sendmail' or $prog eq 'sm-mta') {
 		if($text =~ /\bmailer=local\b/ ) {
@@ -853,6 +858,11 @@
 			event($time, 'virus');
 		}
 	}
+	elsif($prog eq 'clapf') {
+		if($text =~ /SPAM/) {
+			event($time, 'spam');
+		}
+	}
 }
 
 sub event($$)
